doc/abyss-tofastq.1
.TH abyss-tofastq "1" "2015-May" "ABySS 2.3.10" "User Commands"
.SH NAME
abyss-tofastq \- convert various file formats to FASTQ format
.br
abyss-tofasta \- convert various file formats to FASTA format
.SH SYNOPSIS
\fBabyss-tofastq\fR [\fIFILE\fR]...
.br
or
.br
\fBabyss-tofasta\fR [\fIFILE\fR]...
.SH DESCRIPTION
Convert the input files to FASTA or FASTQ format. FILE may be in
FASTA, FASTQ, qseq, export, SRA, SAM or BAM format and may be
compressed with gz, bz2 or xz and may be tarred.
.SH AUTHOR
Written by Shaun Jackman.
.SH "REPORTING BUGS"
Report bugs to <abyss-users@bcgsc.ca>.
.SH COPYRIGHT
Copyright 2015 Canada's Michael Smith Genome Sciences Centre


ChangeLog
2024-10-18 Lauren Coombe <lcoombe@bcgsc.ca>

  * Release version 2.3.10

      General:
	    * Fix usage of `N` and `S` parameters for scaffolding stage

2024-09-11 Johnathan Wong <jowong@bcgsc.ca>

  * Release version 2.3.9

      General:
                * Deallocate memory used by alignment data structure in dialign

2024-07-18 Lauren Coombe <lcoombe@bcgsc.ca>

  * Release version 2.3.8
  
    General:
	  * Type changes to allow building with LTO

2023-05-16 Lauren Coombe <lcoombe@bcgsc.ca>

  * Release version 2.3.7

    General:
	* Fix build on macOS-12 and macOS-13 (Thanks @parham-k!)

2023-05-02 Lauren Coombe <lcoombe@bcgsc.ca>

  * Release version 2.3.6

    General:
    * btllib is now a dependency
	* Documentation updates

	abyss-rresolver-short:
	* btllib source code removed due to btllib now being a dependency

2022-05-11 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.5

    General:
	* Fixed compile errors when using the -DNDEBUG flag.
	* Removed deprecated C++ features.


2021-12-20 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.4

    General:
	* Dropped support for gcc5 due to lack of support for more recent C++ features.

	abyss-rresolver-short:
	* Reduced memory consumption.
	* Better calculation of read size proportions in the input dataset.
	* Increased max read read size allowed.

2021-11-10 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.3

	General:
	* Fixed a contig overlap distance assertion.

	abyss-rresolver-short:
	* Multiple read sizes are better handled. E.g. 150 and 151 are processed as one read size.
    * Updated default parameters.

2021-10-13 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.2

	General:
	* Updated the `m` parameter and the `S` and `N` scaffolding parameters to better defaults.

	abyss-rresolver-short:
  * Updated btllib to 1.1.7

2021-04-21 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.1

	abyss-rresolver-short:
  * Fixed a bug that crashed ABySS if Bloom filter size is not specified.
  * Fixed a compilation error for clang-10.

2021-03-22 Vladimir Nikolic <vnikolic@bcgsc.ca>

  * Release version 2.3.0

	General:
	* The RResolver algorithm is now part of the default pipeline. RResolver resolves repeats using short reads and generates assemblies with better contiguity and often fewer misassemblies.

2020-09-17  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.5

	General:
	* Resolve various compilation errors in newer versions of clang
	* Use ntHash's 0th hash as the default hash instead of the 1st hash
	* Added optional RResolver module, not currently part of the ABySS assembly pipeline

	abyss-rresolver-short:
	* Improves genome assemblies at unitig stage by using a sliding window at read size level
	across repeats to determine which paths are correct
	* For further information: consult http://www.birollab.ca/assets/posts/195_Nikolic_Vladimir_HiTSeq_ISMB2020.pdf
	
	misc:
	* Extract all tags in a SAM file

2020-01-30  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.4

	General:
	* Refactor deprecated functions in clang-8

	Sealer:
	* Remove unsupported -D option from help page

	abyss-bloom:
	* Add counting Bloom Filter instruction to help page

	abyss-bloom-dbg:
	* Report coverage information of unitigs

2019-09-20  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.3
	* Revert memory consumption of Bloom filters to pre 2.2.0 behaviour
	ABySS will now share the specified memory among all Bloom filters
	instead of just the counting Bloom filter.
	* Fix gcc-9 compilation warnings

2019-08-16  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.2
	* Fix abyss-overlap for 32-bit systems

2019-08-12  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.1
	* Fix abyss-bloom for macOS

2019-08-01  Johnathan Wong <jowong@bcgsc.ca>

	* Release version 2.2.0
	* Construct a counting Bloom filter instead of a cascading Bloom filter.

	abyss-bloom:
	* Add 'counting' as valid argument to '-t' option to build a counting
	bloom filter

2018-12-04  Sauparna Palchowdhury <spchowdhury@bcgsc.ca>

	* Release version 2.1.5
	* Compiler fixes and increse stack size limits to avoid stack overflows.

	abyss-pe:
	* Add 'ulimit' statements to the Makefile to increse a thread's
	stack size to 64MB.

2018-11-09  Ben Vandervalk <benv@bcgsc.ca>

	* Release version 2.1.4
	* Major improvements to Bloom filter assembly contiguity and
	correctness. Bloom filter assemblies now have equivalent scaffold
	contiguity and better correctness than MPI assemblies of the same
	data, while still requiring less than 1/10th of the memory. On
	human, Bloom filter assembly times are still a few hours longer
	than MPI assemblies (e.g. 17 hours vs. 13 hours, using 48
	threads).

	abyss-pe:
	* Change default value of `m` from 50 => 0, which has the effect
	of disallowing sequence overlaps < k-1 bp. QUAST tests on E. coli /
	C. elegans / H. sapiens showed that both contiguity and
	correctness were improved by allowing only overlaps of k-1 bp
	between sequence ends.

2018-11-05  Ben Vandervalk <benv@bcgsc.ca>

	* Release version 2.1.3
	* Bug fix release

	abyss-bloom:
	* Added `graph` command for visualizing neighbourhoods of the
	Bloom filter de Bruijn graph (produces GraphViz)

	abyss-fixmate-ssq:
	* Fixed missing tab in SAM output which broke ABySS linked reads
	pipeline (Tigmint/ARCS)

2018-10-24  Ben Vandervalk <benv@bcgsc.ca>

	* Release version 2.1.2
	* Improved scaffold N50 on human by ~10% by
	implementing a new `DistanceEst --median`
	option (thanks to @lcoombe!)
	* Added a `--max-cost` option to `konnector` and
	`abyss-sealer`. This address a long-standing issue
	with indeterminately long running times, particularly
	for low values of `k`.

	abyss-pe:
	* Use the new `DistanceEst --median` option as the
	default for the scaffolding stage

	Dockerfile:
	* Fix OpenMPI setup

	DistanceEst:
	* Added `--median` option

	konnector:
	* Added `--max-cost` option to bound running time


DataLayer/abyss-tofastq.cc
/** Convert various file formats to FASTQ format.
 * Written by Shaun Jackman <sjackman@bcgsc.ca>.
 */
#include "config.h"
#include "Common/IOUtil.h"
#include "Common/Uncompress.h"
#include "DataLayer/FastaInterleave.h"
#include "DataLayer/FastaReader.h"
#include "DataLayer/Options.h"
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <getopt.h>
#include <iostream>
#include <sstream>

using namespace std;

#define PROGRAM "abyss-tofastq"

static const char VERSION_MESSAGE[] =
PROGRAM " (" PACKAGE_NAME ") " VERSION "\n"
"Written by Shaun Jackman.\n"
"\n"
"Copyright 2014 Canada's Michael Smith Genome Sciences Centre\n";

static const char USAGE_MESSAGE[] =
"Usage: " PROGRAM " [OPTION]... [FILE]...\n"
"Convert to FASTQ format. The input format may be FASTA, FASTQ,\n"
"qseq, export, SAM or BAM format and compressed with gz, bz2 or xz\n"
"and may be tarred.\n"
"\n"
" Options:\n"
"\n"
"      --cat               concatenate the records [default]\n"
"  -i, --interleave        interleave the records\n"
"      --fastq             output FASTQ format [default]\n"
"      --fasta             output FASTA format\n"
"      --bx                copy SAM BX tag to output FASTA comment\n"
"      --chastity          discard unchaste reads [default]\n"
"      --no-chastity       do not discard unchaste reads\n"
"      --trim-masked       trim masked bases from the ends of reads\n"
"      --no-trim-masked    do not trim masked bases from the ends\n"
"                          of reads [default]\n"
"  -q, --trim-quality=N    trim bases from the ends of reads whose\n"
"                          quality is less than the threshold\n"
"      --standard-quality  zero quality is `!' (33)\n"
"                          default for FASTQ and SAM files\n"
"      --illumina-quality  zero quality is `@' (64)\n"
"                          default for qseq and export files\n"
"  -v, --verbose           display verbose output\n"
"      --help              display this help and exit\n"
"      --version           output version information and exit\n"
"\n"
"Report bugs to <" PACKAGE_BUGREPORT ">.\n";

namespace opt {
	/** Interleave the lines of each file. */
	static int interleave;

	static int toFASTQ = 1;
	static int verbose;
}

static const char shortopts[] = "iq:v";

enum { OPT_HELP = 1, OPT_VERSION, OPT_BX };

static const struct option longopts[] = {
	{ "cat",              no_argument, &opt::interleave, 0 },
	{ "interleave",       no_argument, &opt::interleave, 1 },
	{ "fasta",            no_argument, &opt::toFASTQ, 0 },
	{ "fastq",            no_argument, &opt::toFASTQ, 1 },
	{ "bx",               no_argument, &opt::bxTag, 1 },
	{ "chastity",         no_argument, &opt::chastityFilter, 1 },
	{ "no-chastity",      no_argument, &opt::chastityFilter, 0 },
	{ "trim-masked",      no_argument, &opt::trimMasked, 1 },
	{ "no-trim-masked",   no_argument, &opt::trimMasked, 0 },
	{ "trim-quality",     required_argument, NULL, 'q' },
	{ "standard-quality", no_argument, &opt::qualityOffset, 33 },
	{ "illumina-quality", no_argument, &opt::qualityOffset, 64 },
	{ "help",             no_argument, NULL, OPT_HELP },
	{ "version",          no_argument, NULL, OPT_VERSION },
	{ NULL, 0, NULL, 0 }
};

/** FastaReader flags. */
static const int FASTAREADER_FLAGS
	= FastaReader::NO_FOLD_CASE | FastaReader::CONVERT_QUALITY;

/** Total count. */
static struct {
	unsigned records;
	unsigned characters;
} g_total;

template <class Record>
static void convert(const char* path)
{
	FastaReader in(path, FASTAREADER_FLAGS);
	unsigned records = 0, characters = 0;
	for (Record record; in >> record;) {
		cout << record;
		assert_good(cout, "stdout");
		records++;
		characters += record.seq.size();
	}
	assert(in.eof());

	g_total.records += records;
	g_total.characters += characters;
	if (opt::verbose)
		cerr << records << '\t'
			<< characters << '\t'
			<< path << '\n';
}

/** Interleave the records. */
template <typename Record>
static void interleave(char** first, char** last)
{
	FastaInterleave in(first, last, FASTAREADER_FLAGS);
	unsigned records = 0, characters = 0;
	for (Record record; in >> record;) {
		cout << record;
		assert_good(cout, "stdout");
		records++;
		characters += record.seq.size();
	}
	assert(in.eof());

	g_total.records += records;
	g_total.characters += characters;
}

int main(int argc, char** argv)
{
	opt::trimMasked = false;

	if (string(argv[0]).find("tofasta") != string::npos)
		opt::toFASTQ = false;

	bool die = false;
	for (int c; (c = getopt_long(argc, argv,
					shortopts, longopts, NULL)) != -1;) {
		istringstream arg(optarg != NULL ? optarg : "");
		switch (c) {
			case '?':
				die = true;
				break;
			case 'i':
				opt::interleave = true;
				break;
			case 'q':
				arg >> opt::qualityThreshold;
				break;
			case 'v':
				opt::verbose++;
				break;
			case OPT_BX:
				opt::bxTag = true;
				break;
			case OPT_HELP:
				cout << USAGE_MESSAGE;
				exit(EXIT_SUCCESS);
			case OPT_VERSION:
				cout << VERSION_MESSAGE;
				exit(EXIT_SUCCESS);
		}
		if (optarg != NULL && !arg.eof()) {
			cerr << PROGRAM ": invalid option: `-"
				<< (char)c << optarg << "'\n";
			exit(EXIT_FAILURE);
		}
	}
	if (die) {
		cerr << "Try `" << PROGRAM
			<< " --help' for more information.\n";
		exit(EXIT_FAILURE);
	}

	typedef void (*F)(const char*);
	F convertFasta = convert<FastaRecord>;
	F convertFastq = convert<FastqRecord>;
	F f = opt::toFASTQ ? convertFastq : convertFasta;

	if (optind == argc) {
		f("-");
	} else if (!opt::interleave || argc - optind == 1) {
		// Concatenate.
		for_each(argv + optind, argv + argc, f);
	} else {
		// Interleave.
		if (opt::toFASTQ)
			interleave<FastqRecord>(argv + optind, argv + argc);
		else
			interleave<FastaRecord>(argv + optind, argv + argc);
	}

	if (opt::verbose && argc - optind > 1)
		cerr << g_total.records << '\t'
			<< g_total.characters << '\t'
			<< "total\n";

	cout.flush();
	assert_good(cout, "stdout");
	return 0;
}
